File: store/container.go
package store

import (
	"bytes"
	"encoding/binary"
	"fmt"

	proposerblock "github.com/ava-labs/avalanchego/vms/proposervm/block"
	"github.com/ava-labs/libevm/core/types"
	"github.com/ava-labs/libevm/rlp"
)

// CONTAINER REASSEMBLY.
//
// epochdb never stores a container whole. It stores hdr/<height> (the header
// RLP), pvm/<height> (everything else about the container) and tx/<txnum> (the
// tx RLPs), and p2p block serving still has to hand a peer back the VERBATIM
// bytes the container was fetched as (DESIGN entry point 5).
//
// THE pvm ROW IS A TEMPLATE, NOT THE WRAPPER BYTES. A container laid out flat
// is:
//
//	A | headerRLP | B | tx1..txN | C
//
// where A is the proposervm wrapper prefix plus the inner block's RLP list
// header, B is the transaction list's RLP header, and C is everything after
// the last transaction: the inner block's remaining fields (uncles, and for
// coreth also Version and ExtData) plus the wrapper's suffix (its signature).
// A, B and C are stored verbatim as
//
//	pvm = [4B big-endian len(A)] A [4B big-endian len(B)] B C
//
// so Reassemble is pure concatenation: no RLP is re-encoded, no list length is
// recomputed (the transactions are stored verbatim, so every length in A and B
// is still correct), and the shape of the block is irrelevant. That last point
// is why this is a template and not "the wrapper prefix and suffix around the
// inner block": coreth's block body carries Version and ExtData AFTER the
// uncles, and a scheme that rebuilds the inner block from header + txs alone
// silently drops them.
//
// An EMPTY pvm row is the shorthand for the common pre-proposervm container:
// no wrapper at all, and an inner block that is exactly the three-field geth
// shape [header, txs, uncles] with no uncles. SplitContainer emits the
// shorthand only when it has checked that it reproduces the container byte for
// byte, so it is never a guess.

// SplitContainer takes a raw container as fetched and returns the pvm row and
// the inner eth block. The libevm extras must already be registered
// (fetch.RegisterExtras).
//
// It CHECKS its own work: the pieces epochdb will store (the re-encoded header
// and transactions, plus the returned pvm row) are reassembled and compared to
// raw before returning. A container that does not round trip is rejected here,
// where it can still be diagnosed, rather than years later when a peer asks
// for it.
func SplitContainer(raw []byte) (pvm []byte, inner *types.Block, err error) {
	if len(raw) == 0 {
		return nil, nil, fmt.Errorf("store: empty container")
	}

	// Same decoder order as fetch/parse.go: proposervm first, then the
	// pre-fork bare RLP block.
	innerBytes := raw
	proposerBlk, perr := proposerblock.ParseWithoutVerification(raw)
	if perr == nil {
		innerBytes = proposerBlk.Block()
	} else {
		_, _, rest, serr := rlp.Split(raw)
		if serr != nil {
			return nil, nil, fmt.Errorf("store: rlp split: %w (not a ProposerVM block either: %v)", serr, perr)
		}
		innerBytes = raw[:len(raw)-len(rest)]
	}

	inner = new(types.Block)
	if err := rlp.DecodeBytes(innerBytes, inner); err != nil {
		return nil, nil, fmt.Errorf("store: decode inner eth block: %w", err)
	}

	headerRLP, txRLPs, err := blockPieces(inner)
	if err != nil {
		return nil, nil, err
	}

	// The shorthand first: if the bare three-field rebuild already is the
	// container, the pvm row is empty.
	if out, e := Reassemble(nil, headerRLP, txRLPs); e == nil && bytes.Equal(out, raw) {
		return nil, inner, nil
	}

	off := bytes.Index(raw, innerBytes)
	if off < 0 {
		return nil, nil, fmt.Errorf("store: inner block bytes are not a substring of the container")
	}
	hs, he, ts, te, err := blockSpans(innerBytes)
	if err != nil {
		return nil, nil, err
	}
	pvm = pvmTemplate(raw, off+hs, off+he, off+ts, off+te)

	out, err := Reassemble(pvm, headerRLP, txRLPs)
	if err != nil {
		return nil, nil, err
	}
	if !bytes.Equal(out, raw) {
		return nil, nil, fmt.Errorf("store: container does not round trip (%d bytes in, %d out)", len(raw), len(out))
	}
	return pvm, inner, nil
}

// pvmTemplate cuts the container into the three template pieces around the
// header span [hs, he) and the transaction-payload span [ts, te), and packs
// them as [4B len(A)] A [4B len(B)] B C.
func pvmTemplate(raw []byte, hs, he, ts, te int) []byte {
	a, b, c := raw[:hs], raw[he:ts], raw[te:]
	pvm := binary.BigEndian.AppendUint32(make([]byte, 0, 8+len(a)+len(b)+len(c)), uint32(len(a)))
	pvm = append(pvm, a...)
	pvm = binary.BigEndian.AppendUint32(pvm, uint32(len(b)))
	pvm = append(pvm, b...)
	return append(pvm, c...)
}

// Reassemble rebuilds the verbatim container from the stored pieces.
func Reassemble(pvm, headerRLP []byte, txRLPs [][]byte) ([]byte, error) {
	if len(pvm) == 0 {
		txs := make([]rlp.RawValue, len(txRLPs))
		for i, t := range txRLPs {
			txs[i] = t
		}
		return rlp.EncodeToBytes(&bareBlock{Header: headerRLP, Txs: txs})
	}
	if len(pvm) < 8 {
		return nil, fmt.Errorf("store: pvm row is %d bytes, too short for a template", len(pvm))
	}
	la := int(binary.BigEndian.Uint32(pvm[:4]))
	if 4+la+4 > len(pvm) {
		return nil, fmt.Errorf("store: pvm row: bad prefix length %d", la)
	}
	a := pvm[4 : 4+la]
	lb := int(binary.BigEndian.Uint32(pvm[4+la : 8+la]))
	if 8+la+lb > len(pvm) {
		return nil, fmt.Errorf("store: pvm row: bad tx-list header length %d", lb)
	}
	b, c := pvm[8+la:8+la+lb], pvm[8+la+lb:]

	n := len(a) + len(headerRLP) + len(b) + len(c)
	for _, t := range txRLPs {
		n += len(t)
	}
	out := make([]byte, 0, n)
	out = append(out, a...)
	out = append(out, headerRLP...)
	out = append(out, b...)
	for _, t := range txRLPs {
		out = append(out, t...)
	}
	return append(out, c...), nil
}

// bareBlock is the pre-proposervm shorthand: the three-field geth block RLP,
// written from bytes epochdb already holds. rlp.RawValue is copied through
// verbatim, so the list headers are the only thing computed here.
type bareBlock struct {
	Header rlp.RawValue
	Txs    []rlp.RawValue
	Uncles []rlp.RawValue
}

// blockPieces re-encodes the rows epochdb stores for a block: hdr/<height> and
// one tx/<txnum> per transaction.
func blockPieces(b *types.Block) (headerRLP []byte, txRLPs [][]byte, err error) {
	headerRLP, err = rlp.EncodeToBytes(b.Header())
	if err != nil {
		return nil, nil, fmt.Errorf("store: encode header: %w", err)
	}
	for i, tx := range b.Transactions() {
		enc, err := rlp.EncodeToBytes(tx)
		if err != nil {
			return nil, nil, fmt.Errorf("store: encode tx %d: %w", i, err)
		}
		txRLPs = append(txRLPs, enc)
	}
	return headerRLP, txRLPs, nil
}

// blockSpans locates the header element and the transaction-list PAYLOAD
// inside an inner eth block RLP, as offsets into it. Everything before hs,
// between he and ts, and after te is template.
func blockSpans(inner []byte) (hs, he, ts, te int, err error) {
	content, rest, err := rlp.SplitList(inner)
	if err != nil {
		return 0, 0, 0, 0, fmt.Errorf("store: block is not an RLP list: %w", err)
	}
	hs = len(inner) - len(rest) - len(content)
	_, _, afterHeader, err := rlp.Split(content)
	if err != nil {
		return 0, 0, 0, 0, fmt.Errorf("store: block header element: %w", err)
	}
	he = len(inner) - len(afterHeader)
	txs, afterTxs, err := rlp.SplitList(afterHeader)
	if err != nil {
		return 0, 0, 0, 0, fmt.Errorf("store: block transaction list: %w", err)
	}
	ts = len(inner) - len(afterTxs) - len(txs)
	return hs, he, ts, ts + len(txs), nil
}

File: rs/store/src/container.rs
//! Container reassembly (store/container.go): the pvm row is a template
//! [4B len(A)] A [4B len(B)] B C around the header RLP and the tx RLPs, or
//! empty for the bare three-field pre-proposervm block.

use anyhow::{anyhow, bail, Result};
use state::rlp;

/// The bare [header, [txs], []] block from stored bytes.
fn bare_block(header_rlp: &[u8], txs: &[&[u8]]) -> Vec<u8> {
    let txs_len: usize = txs.iter().map(|t| t.len()).sum();
    let inner_len = header_rlp.len() + rlp::header_len(txs_len) + txs_len + 1;
    let mut out = Vec::with_capacity(rlp::header_len(inner_len) + inner_len);
    rlp::put_header(&mut out, true, inner_len);
    out.extend_from_slice(header_rlp);
    rlp::put_header(&mut out, true, txs_len);
    for t in txs {
        out.extend_from_slice(t);
    }
    out.push(0xc0);
    out
}

pub fn reassemble(pvm: &[u8], header_rlp: &[u8], txs: &[&[u8]]) -> Result<Vec<u8>> {
    if pvm.is_empty() {
        return Ok(bare_block(header_rlp, txs));
    }
    if pvm.len() < 8 {
        bail!("store: pvm row is {} bytes, too short for a template", pvm.len());
    }
    let la = u32::from_be_bytes(pvm[..4].try_into().unwrap()) as usize;
    if 4 + la + 4 > pvm.len() {
        bail!("store: pvm row: bad prefix length {la}");
    }
    let a = &pvm[4..4 + la];
    let lb = u32::from_be_bytes(pvm[4 + la..8 + la].try_into().unwrap()) as usize;
    if 8 + la + lb > pvm.len() {
        bail!("store: pvm row: bad tx-list header length {lb}");
    }
    let (b, c) = (&pvm[8 + la..8 + la + lb], &pvm[8 + la + lb..]);
    let mut out = Vec::with_capacity(pvm.len() + header_rlp.len() + txs.iter().map(|t| t.len()).sum::<usize>());
    out.extend_from_slice(a);
    out.extend_from_slice(header_rlp);
    out.extend_from_slice(b);
    for t in txs {
        out.extend_from_slice(t);
    }
    out.extend_from_slice(c);
    Ok(out)
}

/// The pvm row for a container whose inner eth block sits at
/// [inner_off, inner_off+inner_len). Checks its own work: the pieces must
/// reassemble to the container byte for byte.
pub fn split_container(raw: &[u8], inner_off: usize, inner_len: usize, header_rlp: &[u8], txs: &[&[u8]]) -> Result<Vec<u8>> {
    if bare_block(header_rlp, txs) == raw {
        return Ok(Vec::new());
    }
    let inner = &raw[inner_off..inner_off + inner_len];
    let (content, rest) = rlp::split_list(inner).ok_or_else(|| anyhow!("store: block is not an RLP list"))?;
    let hs = inner.len() - rest.len() - content.len();
    let (_, _, after_header) = rlp::split(content).ok_or_else(|| anyhow!("store: block header element"))?;
    let he = inner.len() - after_header.len();
    let (txs_content, after_txs) = rlp::split_list(after_header).ok_or_else(|| anyhow!("store: block transaction list"))?;
    let ts = inner.len() - after_txs.len() - txs_content.len();
    let te = ts + txs_content.len();
    let (hs, he, ts, te) = (inner_off + hs, inner_off + he, inner_off + ts, inner_off + te);
    let (a, b, c) = (&raw[..hs], &raw[he..ts], &raw[te..]);
    let mut pvm = Vec::with_capacity(8 + a.len() + b.len() + c.len());
    pvm.extend_from_slice(&(a.len() as u32).to_be_bytes());
    pvm.extend_from_slice(a);
    pvm.extend_from_slice(&(b.len() as u32).to_be_bytes());
    pvm.extend_from_slice(b);
    pvm.extend_from_slice(c);
    let out = reassemble(&pvm, header_rlp, txs)?;
    if out != raw {
        bail!("store: container does not round trip ({} bytes in, {} out)", raw.len(), out.len());
    }
    Ok(pvm)
}

/// The transaction elements of an inner eth block, each as its full RLP
/// bytes (a typed tx keeps its string header): what a tx/ row stores.
pub fn tx_elements(inner: &[u8]) -> Result<Vec<&[u8]>> {
    let (content, _) = rlp::split_list(inner).ok_or_else(|| anyhow!("store: block is not an RLP list"))?;
    let (_, _, after_header) = rlp::split(content).ok_or_else(|| anyhow!("store: block header element"))?;
    let (mut txs, _) = rlp::split_list(after_header).ok_or_else(|| anyhow!("store: block transaction list"))?;
    let mut out = Vec::new();
    while !txs.is_empty() {
        let (_, _, rest) = rlp::split(txs).ok_or_else(|| anyhow!("store: transaction element"))?;
        out.push(&txs[..txs.len() - rest.len()]);
        txs = rest;
    }
    Ok(out)
}

/// The envelope of a stored tx element (eth_getRawTransaction's bytes,
/// what the tx hash is over): the element itself for a legacy tx, the
/// string's content for a typed one.
pub fn tx_envelope(elem: &[u8]) -> &[u8] {
    if elem.first().map(|b| *b >= 0xc0).unwrap_or(true) {
        return elem;
    }
    rlp::split_string(elem).map(|(c, _)| c).unwrap_or(elem)
}
